### PR TITLE
Make the wellModel self-contained

### PR DESCRIPTION
**Pull request by totto82**
The wellModel is now persistent over the time steps,
with an update method called every reportStep/episode.

This allows the following simplifications:
1) move the wellState to the WellModel
2) add a ref to the ebosSimulator to the wellModel
3) clean up the parameters passed to the wellModel methods
4) split the legacy ADB VFP table code.
5) move RESV handling to the WellModel and the rateConverter
6) move the econLimit update to the WellModel

This has been tested on norne and with unit tests on OPM simulators.

This should not be merged before the release. 
It also needs to be tested on Rel0++

### CODE REVIEW

**Comment by totto82** [comment]
jenkins build this please

**Comment by andlaus** [comment]
jenkins fails because `WellSwitchingLogger.hpp` is located in `opm/simulators` instead of `opm/autodiff`. This is probably because you've copied the file on your machine and forgot to remove the one in the old location.

**Comment by totto82** [comment]
This PR was not suppose to have an impact on the run, but it does. The good news is that it improves the convergence on all tested cases: Rel0, Rel5 and Norne. Still I think we need to understand why before this is ready to be merged.

**Comment by totto82** [comment]
> This PR was not suppose to have an impact on the run, but it does. The good news is that it improves the convergence on all tested cases: Rel0, Rel5 and Norne. Still I think we need to understand why before this is ready to be merged.

For some reason changing the call sequence of the  wellModel().apply(ebosResid) created some disturbance in the non-linear solution path. I moved this back to where it was in master. 
I plan to make a new PR that moves this to the solveJacobianSystem() call where I think it conceptually belongs.

**Comment by totto82** [comment]
jenkins build this please

File: opm/autodiff/BlackoilWellModel_impl.hpp [hunk 1]
```diff
@@ -205,13 +237,11 @@ namespace Opm {
...
-    assembleWellEq(Simulator& ebosSimulator,
-                   const double dt,
-                   WellState& well_state,
-                   bool only_wells) const
+    assembleWellEq(const double dt,
+                   bool only_wells)
     {
-        for (int w = 0; w < number_of_wells_; ++w) {
-            well_container_[w]->assembleWellEq(ebosSimulator, dt, well_state, only_wells);
+        for (int w = 0; w < wells()->number_of_wells; ++w) {
```
Comment (GitPaean) [inline opm/autodiff/BlackoilWellModel_impl.hpp]: The previous efforts were made to only use `Wells` struct in the initialization, maybe also one place related to group control. The point is that when we want to remove the `Wells` struct, we can do that relatively easily. I think it was something people like to have. Maybe in this PR due to WellState, we need a few more places to use it, but please do not undo other related efforts.

Comment (GitPaean) [reply opm/autodiff/BlackoilWellModel_impl.hpp]: now I saw `Wells` struct was also used in `resetWellControlFromState`, it is something can be changed to use the `well_container_`.

File: opm/autodiff/BlackoilWellModel_impl.hpp [hunk 2]
```diff
@@ -5,139 +5,167 @@ namespace Opm {
...
     {
-        // has to be set always for the convergence check!
-        global_nc_   = global_nc;
+        const Grid& grid = ebosSimulator_.gridManager().grid();
+        const auto& defunct_well_names = ebosSimulator_.gridManager().defunctWellNames();
+        const auto& eclState = ebosSimulator_.gridManager().eclState();
+        wells_ecl_ = eclState.getSchedule().getWells(timeStepIdx);
+
+        // Create wells and well state.
+        wells_manager_.reset( new WellsManager (eclState,
```
Comment (GitPaean) [inline opm/autodiff/BlackoilWellModel_impl.hpp]: `WellsManager` is created each report step, it does not need to be a member. `WellCollection` can be a smaller member to keep.

**Comment by GitPaean** [comment]
> For some reason changing the call sequence of the wellModel().apply(ebosResid) created some disturbance in the non-linear solution path.

I did not see where you changed the call sequence. Any more information there?

File: opm/autodiff/WellInterface.hpp [hunk 3]
```diff
@@ -64,6 +63,11 @@ namespace Opm
         using WellState = WellStateFullyImplicitBlackoil;
 
         typedef BlackoilModelParameters ModelParameters;
+
+        static const int Water = BlackoilPhases::Aqua;
```
Comment (GitPaean) [inline opm/autodiff/WellInterface.hpp]: Can we use `Water`, `Oil` and `Gas` also for other places instead of `BlackoilPhases::`?

File: opm/autodiff/BlackoilWellModel_impl.hpp (hunk 1 above)
Comment (andlaus) [reply opm/autodiff/BlackoilWellModel_impl.hpp]: The point of this change is to reduce the amount of redundant -- and thus potentially inconsistent -- information. I'm not aware of anybody who likes `struct Wells`, but with this PR it becomes an completely internal implementation detail of `WellModel`, so the should not harm any removal efforts?

File: opm/autodiff/BlackoilWellModel_impl.hpp (hunk 2 above)
Comment (andlaus) [reply opm/autodiff/BlackoilWellModel_impl.hpp]: after this PR, the well model object is persistent for the whole simulation. this is done in order to be able to fold the `WellState`/`struct Wells`/`WellManager`/`WellCollection`/`RateConverter`/`VFP` logic into the `WellModel` class so that all of this stuff can be refactored/removed without touching outside code. this is the primary semantic change of this PR ;)

File: opm/autodiff/BlackoilWellModel_impl.hpp (hunk 1 above)
Comment (GitPaean) [reply opm/autodiff/BlackoilWellModel_impl.hpp]: To avoid inconsistency, I will suggest 
```
int numWells() const
{
    return wells()->number_of_wells;
}

int numPhases() const
{
    return wells()->number_of_phases;
}
```
We have store the WellControls for each individual well, it looks like these are the only information we need from the `Wells struct`.

**Comment by andlaus** [comment]
> I did not see where you changed the call sequence. Any more information there?

the problem was that the residual part of the Schur complement was applied in `solveJacobianSystem()` instead of in `assemble()` which caused the `getConvergence()` method to behave differently. (`getConvergence()` is called between assembly and linear solve.) not including the Schur  complement in the convergence check is the right thing to do, though.

File: opm/autodiff/WellInterface.hpp (hunk 3 above)
Comment (andlaus) [reply opm/autodiff/WellInterface.hpp]: maybe it is better to go the whole way and switch this to `FluidSystem::{water,gas,oil}PhaseIdx`...

Comment (GitPaean) [reply opm/autodiff/WellInterface.hpp]: All these indices causing confusions from time to time. It is good to stick to only one when possible. I do not have a preference on which one is better than others.

Comment (andlaus) [reply opm/autodiff/WellInterface.hpp]: in all the eWoms parts, only the of the names from the `FluidSystem` are available, so we should probably also use those here. (As I see it, the other phase names are only used for historical reasons.)

File: opm/autodiff/BlackoilWellModel_impl.hpp (hunk 2 above)
Comment (GitPaean) [reply opm/autodiff/BlackoilWellModel_impl.hpp]: Sure. It is just nitpickking for discussion. I see now `Wells struct` is also from `wells_manager_`.

Comment (andlaus) [reply opm/autodiff/BlackoilWellModel_impl.hpp]: > Sure. It is just nitpickking for discussion.

That's how I interpreted you. Thanks for having a thorough look, BTW; sometimes it is hard to see such issues in changes if they are too familiar.

**Comment by GitPaean** [comment]
> the problem was that the residual part of the Schur complement was applied in solveJacobianSystem() instead of in assemble() which caused the getConvergence() method to behave differently. 

You mean `was` in this PR not in master branch, right?  I did not notice that. Thanks for the explanation. 

> not including the Schur complement in the convergence check is the right thing to do, though.
I agree with this.

File: opm/autodiff/WellInterface.hpp (hunk 3 above)
Comment (totto82) [reply opm/autodiff/WellInterface.hpp]: I think it is best to stick to BlackoilPhases:: for the moment since that's what is used in the rest of the well model. I couldn't find any other place where BlackoilPhases is used in the code? Could you point me to where it is?

File: opm/autodiff/BlackoilWellModel_impl.hpp (hunk 1 above)
Comment (totto82) [reply opm/autodiff/BlackoilWellModel_impl.hpp]: I added numWells() and numPhases() to minimize the usage of wells(). Removing all traces of wells() is for another PR and day.

**Comment by totto82** [comment]
jenkins build this please

File: opm/autodiff/WellInterface.hpp (hunk 3 above)
Comment (GitPaean) [reply opm/autodiff/WellInterface.hpp]: Thanks for the response. Now I can see you moved same definition of `Water`, `Oil` and `Gas` out of the `BlackoilEnums`.  I always get confused by the `Water`, `Oil` and `Gas` and their `BlackoilPhases:` ones.  I can see now the efforts of removing `BlackoilEnums` from Ebos in this PR. It is Okay. Sorry for the noise.

**Comment by totto82** [comment]
jenkins build this please

File: opm/autodiff/WellDensitySegmented.cpp [hunk 4]
```diff
@@ -17,10 +17,10 @@
   along with OPM.  If not, see <http://www.gnu.org/licenses/>.
 */
 
-#include <opm/autodiff/WellDensitySegmented.hpp>
+#include "WellDensitySegmented.hpp"
```
Comment (atgeirr) [inline opm/autodiff/WellDensitySegmented.cpp]: Please use the standard include notation with full path, even for headers in the same directory. (This probably applies to several other places as well.)

File: opm/autodiff/WellDensitySegmented.cpp [hunk 5]
```diff
@@ -150,124 +150,6 @@ Opm::WellDensitySegmented::computeConnectionDensities(const Wells& wells,
 
 
 std::vector<double>
-Opm::WellDensitySegmented::computeConnectionDensities(const Wells& wells,
```
Comment (atgeirr) [inline opm/autodiff/WellDensitySegmented.cpp]: Also remove from header file.

File: opm/autodiff/FlowMainEbos.hpp [hunk 6]
```diff
@@ -728,8 +728,7 @@ namespace Opm
                                            FluidSystem::enableDissolvedGas(),
                                            FluidSystem::enableVaporizedOil(),
                                            eclState(),
-                                           *output_writer_,
-                                           defunctWellNames()));
```
Comment (atgeirr) [inline opm/autodiff/FlowMainEbos.hpp]: Also remove the defunctWellNames() member, it does not seem to be called anymore.

File: opm/autodiff/MultisegmentWell_impl.hpp [hunk 7]
```diff
@@ -19,7 +19,7 @@
 */
 
 
-#include <opm/autodiff/MSWellHelpers.hpp>
+#include "MSWellHelpers.hpp"
```
Comment (atgeirr) [inline opm/autodiff/MultisegmentWell_impl.hpp]: Since there seems to be a pattern here: did you experience any problems related to include paths?

File: opm/autodiff/MultisegmentWells.hpp [hunk 8]
```diff
@@ -237,6 +237,10 @@ namespace Opm {
 
             const Vector& wellPerfEfficiencyFactors() const;
 
+            // just return the passed well state
+            template<class WellState>
+            const WellState& wellState(const WellState& well_state) const { return well_state; }
```
Comment (atgeirr) [inline opm/autodiff/MultisegmentWells.hpp]: I do not understand the use of this function. Could you explain?

File: opm/autodiff/RateConverter.hpp [hunk 9]
```diff
@@ -21,13 +21,12 @@
...
 #define OPM_RATECONVERTER_HPP_HEADER_INCLUDED
 
-#include <opm/autodiff/BlackoilPropsAdFromDeck.hpp>
-
 #include <opm/core/props/BlackoilPhases.hpp>
 #include <opm/core/simulator/BlackoilState.hpp>
 #include <opm/core/utility/RegionMapping.hpp>
 #include <opm/core/linalg/ParallelIstlInformation.hpp>
 
+#include <dune/grid/common/gridenums.hh>
```
Comment (atgeirr) [inline opm/autodiff/RateConverter.hpp]: Why is this needed?

File: opm/autodiff/BlackoilWellModel.hpp [hunk 10]
```diff
@@ -69,6 +69,10 @@ namespace Opm {
             typedef WellStateFullyImplicitBlackoil WellState;
             typedef BlackoilModelParameters ModelParameters;
 
+            static const int Water = BlackoilPhases::Aqua;
```
Comment (atgeirr) [inline opm/autodiff/BlackoilWellModel.hpp]: I suggest getting these from WellInterface instead, so there is only one definition instead of a duplication.

File: opm/autodiff/WellStateFullyImplicitBlackoil.hpp [hunk 11]
```diff
@@ -35,6 +34,7 @@
 #include <map>
 #include <algorithm>
 #include <array>
+#include <iostream>
```
Comment (atgeirr) [inline opm/autodiff/WellStateFullyImplicitBlackoil.hpp]: Debug include?

File: tests/test_vfpproperties.cpp [hunk 12]
```diff
@@ -46,9 +46,9 @@
 #include <opm/parser/eclipse/EclipseState/Tables/VFPProdTable.hpp>
 #include <opm/parser/eclipse/Units/UnitSystem.hpp>
 
-#include <opm/autodiff/VFPProperties.hpp>
```
Comment (atgeirr) [inline tests/test_vfpproperties.cpp]: Would it not be better to have the tests deal with the non-ADB version of the VFP code, as that is more likely to live long and prosper?

File: opm/autodiff/WellDensitySegmented.cpp (hunk 4 above)
Comment (andlaus) [reply opm/autodiff/WellDensitySegmented.cpp]: the point of this is that with the quotes, these files can be moved to another directory without having to change them. (we thought about moving them to a `wells/` subdirectory.)

File: opm/autodiff/MultisegmentWells.hpp (hunk 8 above)
Comment (andlaus) [reply opm/autodiff/MultisegmentWells.hpp]: this is currently only needed for restart. Primarily because we did not want to open that can of worms in the same PR.

File: opm/autodiff/RateConverter.hpp (hunk 9 above)
Comment (andlaus) [reply opm/autodiff/RateConverter.hpp]: without it, `RateConverter.hpp` is not autonomous enough to make `test_rateconverter.cpp` compile: it will produce an error for https://github.com/OPM/opm-simulators/blob/master/opm/autodiff/RateConverter.hpp#L500.

Comment (andlaus) [reply opm/autodiff/RateConverter.hpp]: (before that file was indirectly included via BlackoilPropsAdFromDeck.hpp)

File: tests/test_vfpproperties.cpp (hunk 12 above)
Comment (andlaus) [reply tests/test_vfpproperties.cpp]: sure. that would be a quite bit more work, though?

File: opm/autodiff/MultisegmentWells.hpp (hunk 8 above)
Comment (atgeirr) [reply opm/autodiff/MultisegmentWells.hpp]: Ok, but what is the pattern here? I see that it's probably related to the small change in the output code, but I don't understand the need for this?

**Comment by totto82** [comment]
> This PR is very large, but it should be fairly easy to split it into more digestible pieces. For example, the splitting of "ADB vs. Evaluation" versions of the VFP things create a lot of code duplication, and shadows the main point of this PR. Is there a reason this change must be done simultaneously with the other changes?

In retrospective, you are right. I didn't expect the VFP things to create so much changes. Do you want me to split it. It will take some time to get the splitting right, so I would appreciate not to split it.

File: tests/test_vfpproperties.cpp (hunk 12 above)
Comment (totto82) [reply tests/test_vfpproperties.cpp]: You are right. We should test the non-ADB code. I will make an update.

File: opm/autodiff/MultisegmentWells.hpp (hunk 8 above)
Comment (totto82) [reply opm/autodiff/MultisegmentWells.hpp]: This is the same trick that is used by the output code. Instead of changing the interface of the solver and timestepping classes we just add a trivial method for the legacy code.

File: opm/autodiff/WellDensitySegmented.cpp (hunk 4 above)
Comment (atgeirr) [reply opm/autodiff/WellDensitySegmented.cpp]: While that may be a minor benefit for some cases where the files can be assumed to always be in the same directory (such as here, where a cpp file includes its hpp file) this PR contains many instances of "quote-inclusion" that are not just including the associated header file. Also, it prevents (without changes) moving headers to separate directories which is done by some libraries (although I certainly prefer to have them in the same directory, I am not proposing to change that!).

In this PR it mostly adds unnecessary noise. I was going to ask you to put that commit in its own PR at least, but then this PR is just a single commit...! (Candidate for biggest non-trivial commit in OPM history?)

Comment (atgeirr) [reply opm/autodiff/WellDensitySegmented.cpp]: To make it clear: I do not want to mess up the code (further) by introducing this quoting style (I am aware that it already snuck in some places, but I have not had the energy to stamp them out) when the existing style is very clear and good. Should you choose to fight for this change, please use a separate PR, and it would take additional arguments to convince me...!

**Comment by atgeirr** [comment]
> It will take some time to get the splitting right, so I would appreciate not to split it.

I don't really see why that should be complicated? Is it not just
 1) all changes to VFP* files, and
 2) include directive changes in the legacy codes

that would go into this? Then we could merge that in a completely separate operation.

**Comment by totto82** [comment]
> I don't really see why that should be complicated? Is it not just
> 
>     all changes to VFP* files, and
>     include directive changes in the legacy codes
> 
> that would go into this? Then we could merge that in a completely separate operation.

OK, I will do it. I will also revert the include "   " style before we spent to much time discussion it so this PR becomes purely about stuffing all well related stuff into the well model.

**Comment by atgeirr** [comment]
> OK, I will do it. I will also revert the include " " style before we spent to much time discussion it so this PR becomes purely about stuffing all well related stuff into the well model.

Thanks a lot!

**Comment by atgeirr** [comment]
One more thing: I will continue reading and commenting on this PR, so when you are updating it, consider opening new PRs, since sometimes PR comments get lost when rebasing/force pushing.

**Comment by totto82** [comment]
I will make a new PR.

File: opm/autodiff/FlowMainEbos.hpp (hunk 6 above)
Comment (totto82) [reply opm/autodiff/FlowMainEbos.hpp]: Fixed in new PR

File: opm/autodiff/WellDensitySegmented.cpp (hunk 5 above)
Comment (totto82) [reply opm/autodiff/WellDensitySegmented.cpp]: Fixed in new PR

File: opm/autodiff/BlackoilWellModel.hpp (hunk 10 above)
Comment (totto82) [reply opm/autodiff/BlackoilWellModel.hpp]: Fixed in new PR

File: opm/autodiff/WellStateFullyImplicitBlackoil.hpp (hunk 11 above)
Comment (totto82) [reply opm/autodiff/WellStateFullyImplicitBlackoil.hpp]: Fixed in new PR

File: tests/test_vfpproperties.cpp (hunk 12 above)
Comment (totto82) [reply tests/test_vfpproperties.cpp]: Fixed in new PR

**Comment by atgeirr** [comment]
I'll close this as the main parts have been merged. I suggest using a new pull request for the VFP things.
